python/ray/data/_internal/execution/operators/zip_operator.py

```diff
@@ -236,11 +237,22 @@ def _zip_one_block(
         # Swap blocks if ordering was inverted during block alignment splitting.
         block, other_block = other_block, block
     # Zip block and other blocks.
-    result = BlockAccessor.for_block(block).zip(other_block)
+    result = _try_zip(block, other_block)
     br = BlockAccessor.for_block(result)
     return result, br.get_metadata(input_files=[], exec_stats=stats.build())
 
 
 def _get_num_rows_and_bytes(block: Block) -> Tuple[int, int]:
     block = BlockAccessor.for_block(block)
     return block.num_rows(), block.size_bytes()
+
+
+def _try_zip(block: Block, other_block: Block) -> Block:
+    try:
+        return BlockAccessor.for_block(block).zip(other_block)
+    except ValueError as e:
+        pd = _lazy_import_pandas()
+        if pd and isinstance(other_block, pd.DataFrame) and hasattr(block, "to_pandas"):
+            return BlockAccessor.for_block(block.to_pandas()).zip(other_block)
+        else:
+            raise "Failed to zip: {}".format(e)
```

Comment:
Can we use arrow blocks as the default fallback type here, instead of pandas? The main motivation being that Arrow blocks typically provide much faster performance (and converting from Arrow to pandas after the fact is usually straightforward).

the high level logic I have in mind is:
- for each block, check if they are Arrow; if not, convert them to Arrow
- try zipping the two (now Arrow) blocks
- if the zip fails, raise a `ValueError`

Reply:
My assumption came from here:

https://github.com/ray-project/ray/blob/08aa1384e8c25f8898a20fd1d647012d1f020400/python/ray/data/_internal/delegating_block_builder.py#L20-L28

it looks like `pyarrow` cannot handle nested Python objects, so I decided the fallback to be pandas.